okhttp/src/commonMain/kotlin/okhttp3/Call.kt

```diff
@@ -62,6 +62,16 @@ expect interface Call {
    */
   fun clone(): Call
 
+  /**
+   * Registers a callback for cancellation of the Call.
+   * Typically used in an interceptor that may be in a blocking
+   * call and may need to proactively cancel other work.
+   *
+   * Most cancellation should be handled via existing
+   * [execute], [enqueue] and executeAsync behaviour.
+   */
+  actual fun onCancel(cancelFn: () -> Unit)
```

Comment:
should this be `Runnable` for our Java language friends?

Reply:
It's in common, I could typealias it, but maybe a single use "fun interface" is better?

Reply:
I think adding a supporting class here is overkill, anyone using this from java can return Unit.Instance, or null.  Or you feel it's important to be a pleasant API?

Reply:
Alternately, would it make sense to let callers provide an entire EventListener here? That gives us and callers lots of power and keeps our API compact?

Reply:
Yep - let's give it a go!

Reply:
It will make this cancellation callback JVM only. EventListener isn't multiplatform friendly (proxy, inetaddress, HttpUrl).

I'll try this and see how it goes.

We can some KMP later, as we get more idea of what it needs to be.

Reply:
we can have something like `chain.withEventListener()` and that can register an additional listener in RealCall.eventListener, which is essentially the event listener holder.

But it becomes complicated.  what happens if the call was cancelled just as your interceptor is run.  do we special case cancel and call end to always be called?